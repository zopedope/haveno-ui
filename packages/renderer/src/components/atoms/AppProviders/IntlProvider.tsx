// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { useEffect, useMemo, useState } from "react";
import { IntlProvider as ReacIntlProvider } from "react-intl";
import type { FC } from "react";
import { LangPack, SupportedLocales } from "@src/constants/lang";
import type { LangKeys } from "@src/constants/lang";

const DEFAULT_LOCALE = SupportedLocales.EN;

export const IntlProvider: FC = ({ children }) => {
  const [messages, setMessages] = useState<{ [key in LangKeys]: string }>();
  const locale = useMemo(
    () =>
      navigator.language in SupportedLocales
        ? navigator.language
        : DEFAULT_LOCALE,
    [navigator.language]
  );

  useEffect(() => {
    setMessages(LangPack[locale] ?? LangPack[DEFAULT_LOCALE]);
  }, [locale]);

  return (
    <ReacIntlProvider locale={locale} messages={messages}>
      {children}
    </ReacIntlProvider>
  );
};
