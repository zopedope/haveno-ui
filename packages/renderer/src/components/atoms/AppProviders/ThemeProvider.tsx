// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { MantineProvider, Global } from "@mantine/core";
import type { FC } from "react";
import { themeOverride, globalStyles } from "@src/theme";

export const ThemeProvider: FC = ({ children }) => {
  return (
    <MantineProvider theme={themeOverride}>
      <Global styles={globalStyles} />
      {children}
    </MantineProvider>
  );
};
