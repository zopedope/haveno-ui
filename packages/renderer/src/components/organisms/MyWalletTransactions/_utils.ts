// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { isEmpty } from "lodash";
import type { XmrTx } from "haveno-ts";
import { WalletTransactionType } from "@molecules/WalletTransactions/_types";
import type { TWalletTransaction } from "@molecules/WalletTransactions/_types";

export const transfromXmrTxs = (
  xmrTxs: Array<XmrTx.AsObject>
): Array<TWalletTransaction> => {
  return xmrTxs.map(transfromXmrTx);
};

const transfromXmrTx = (xmrTx: XmrTx.AsObject): TWalletTransaction => ({
  type: isEmpty(xmrTx.incomingTransfersList)
    ? WalletTransactionType.Sent
    : WalletTransactionType.Received,

  fee: parseFloat(xmrTx.fee),
  height: xmrTx.height,

  amount: !isEmpty(xmrTx.outgoingTransfer)
    ? parseFloat(xmrTx.outgoingTransfer?.amount || "")
    : parseFloat(xmrTx.incomingTransfersList[0]?.amount),

  amountCurrency: "XMR",
  transactionId: xmrTx.hash,

  timestamp: xmrTx.timestamp,

  incomingAddresses: xmrTx.incomingTransfersList?.map((addr) => addr.address),
  destinationAddresses: xmrTx.outgoingTransfer?.destinationsList?.map(
    (addr) => addr.address
  ),
});
