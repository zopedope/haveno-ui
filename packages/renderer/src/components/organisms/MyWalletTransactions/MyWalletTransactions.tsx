// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { useMemo } from "react";
import { Group, Loader } from "@mantine/core";
import type { FC } from "react";
import { transfromXmrTxs } from "./_utils";
import { useXmrTxs } from "@hooks/haveno/useXmrTxs";
import { WalletTransactions } from "@molecules/WalletTransactions";

export function MyWalletTransactionsTable() {
  const { data: xmrTxs } = useXmrTxs();

  const transactions = useMemo(() => transfromXmrTxs(xmrTxs || []), [xmrTxs]);

  return xmrTxs ? <WalletTransactions data={transactions} /> : null;
}

const MyWalletTransactionsBoot: FC = ({ children }) => {
  const { isLoading: isXmrTxsLoading } = useXmrTxs();

  return isXmrTxsLoading ? (
    <Group position="center" pt="lg" pb="lg">
      <Loader color="gray" size="sm" />
    </Group>
  ) : (
    <>{children}</>
  );
};

export function MyWalletTransactions() {
  return (
    <MyWalletTransactionsBoot>
      <MyWalletTransactionsTable />
    </MyWalletTransactionsBoot>
  );
}
