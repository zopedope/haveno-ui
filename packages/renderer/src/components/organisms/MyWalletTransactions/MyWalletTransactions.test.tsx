// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { describe, expect, it, vi, beforeAll } from "vitest";
import { render, screen } from "@testing-library/react";
import { MyWalletTransactions } from "./MyWalletTransactions";
import { AppProviders } from "@atoms/AppProviders";

describe("organisms::MyWalletMoneroBalance", () => {
  beforeAll(() => {
    vi.mock("@hooks/haveno/useXmrTxs", () => ({
      useXmrTxs: () => ({
        isLoading: false,
        isSuccess: true,
        data: [
          {
            timestamp: 1653593643913,
            height: "1.334423",
            fee: "3.33",
            isConfirmed: true,
            isLocked: false,
            hash: "HASHADDRESS",
            incomingTransfersList: [
              {
                amount: "10000",
                accountIndex: 1,
                subaddressIndex: 1,
                address: "INCOMINGADDRESS",
                numSuggestedConfirmations: 100,
              },
            ],
            outgoingTransfer: {},
            metadata: "",
          },
        ],
      }),
    }));
  });

  it("renders without exploding.", () => {
    const { asFragment, unmount } = render(
      <AppProviders>
        <MyWalletTransactions />
      </AppProviders>
    );
    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  it("renders date detail correctly.", () => {
    const { unmount } = render(
      <AppProviders>
        <MyWalletTransactions />
      </AppProviders>
    );
    expect(screen.queryByText("May 26, 2022")).toBeInTheDocument();
    unmount();
  });

  it("renders time detail correctly.", () => {
    const { unmount } = render(
      <AppProviders>
        <MyWalletTransactions />
      </AppProviders>
    );
    expect(screen.queryByText("7:34 PM")).toBeInTheDocument();
    unmount();
  });
});
