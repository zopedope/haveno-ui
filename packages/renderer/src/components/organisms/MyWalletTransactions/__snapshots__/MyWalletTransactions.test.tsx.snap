// Vitest Snapshot v1

exports[`organisms::MyWalletMoneroBalance > renders without exploding. 1`] = `
<DocumentFragment>
  <table
    class="mantine-Table-root mantine-qaazp5"
  >
    <tbody>
      <tr>
        <td
          style="width: 150px;"
        >
          <div
            class="mantine-Group-root mantine-6y1794"
          >
            <div
              class="mantine-CircleIcon-root mantine-Group-child mantine-lh1mzd"
            >
              <svg
                color="#75B377"
                fill="none"
                height="18"
                viewBox="0 0 14 16"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7.83401 14.9637L12.733 10.4695C13.1844 10.0554 13.1844 9.36518 12.733 8.95105C12.2815 8.53691 11.5291 8.53691 11.0777 8.95105L8.16841 11.6046L8.16841 1.78804C8.16841 1.18984 7.65009 0.714355 6.99802 0.714355C6.34594 0.714355 5.82762 1.18984 5.82762 1.78804L5.82762 11.6046L2.93507 8.95105C2.48363 8.53691 1.73124 8.53691 1.2798 8.95105C1.04572 9.16578 0.928677 9.44187 0.928677 9.71796C0.928677 9.99405 1.04572 10.2701 1.2798 10.4849L6.16202 14.9637C6.37938 15.1631 6.68034 15.2858 6.99801 15.2858C7.31569 15.2858 7.61665 15.1784 7.83401 14.9637Z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="mantine-Group-child mantine-1oprzqz"
            >
              <div
                class="mantine-Text-root mantine-8y7ope"
              >
                Received
              </div>
              <div
                class="mantine-Group-root mantine-bhsibn"
              >
                <div
                  class="mantine-Text-root mantine-Group-child mantine-8uy9kb"
                >
                  7:34 PM
                </div>
                <div
                  class="mantine-Text-root mantine-Group-child mantine-8uy9kb"
                >
                  May 26, 2022
                </div>
              </div>
            </div>
          </div>
        </td>
        <td
          style="width: 400px;"
        >
          <div
            class="mantine-Stack-root mantine-1yjxaby"
          >
            <div
              class="mantine-Text-root mantine-8y7ope"
            >
              XMRÂ 10,000.00
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
